File: server/src/trips.ts
import { db } from "./db";

export interface TripRow {
  id: number;
  user_id: number;
  started_at: string;
  ended_at: string | null;
  initial_fuel_liters: number | null;
  final_fuel_liters: number | null;
  total_distance_km: number;
}

export interface TripPointRow {
  id: number;
  trip_id: number;
  timestamp: string;
  lat: number;
  lng: number;
}

// Haversine distance (km)
export function haversineKm(
  lat1: number,
  lon1: number,
  lat2: number,
  lon2: number
): number {
  const R = 6371; // km
  const toRad = (deg: number) => (deg * Math.PI) / 180;
  const dLat = toRad(lat2 - lat1);
  const dLon = toRad(lon2 - lon1);
  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos(toRad(lat1)) *
      Math.cos(toRad(lat2)) *
      Math.sin(dLon / 2) *
      Math.sin(dLon / 2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return R * c;
}

export function getActiveTrip(userId: number): Promise<TripRow | null> {
  return new Promise((resolve, reject) => {
    db.get(
      "SELECT * FROM trips WHERE user_id = ? AND ended_at IS NULL ORDER BY started_at DESC LIMIT 1",
      [userId],
      (err, row: any) => {
        if (err) return reject(err);
        resolve(row ? (row as TripRow) : null);
      }
    );
  });
}

export function startTrip(
  userId: number,
  initialFuel?: number
): Promise<TripRow> {
  return new Promise((resolve, reject) => {
    const stmt = db.prepare(
      "INSERT INTO trips (user_id, initial_fuel_liters) VALUES (?, ?)"
    );
    stmt.run(
      userId,
      initialFuel ?? null,
      function (this: any, err: Error | null) {
        if (err) return reject(err);
        db.get("SELECT * FROM trips WHERE id = ?", [this.lastID], (e, row) => {
          if (e) return reject(e);
          resolve(row as TripRow);
        });
      }
    );
  });
}

export async function addPoint(
  tripId: number,
  lat: number,
  lng: number
): Promise<{ point: TripPointRow; distanceAdded: number; total: number }> {
  // get last point
  const lastPoint: TripPointRow | null = await new Promise(
    (resolve, reject) => {
      db.get(
        "SELECT * FROM trip_points WHERE trip_id = ? ORDER BY timestamp DESC LIMIT 1",
        [tripId],
        (err, row: any) => {
          if (err) return reject(err);
          resolve(row ? (row as TripPointRow) : null);
        }
      );
    }
  );
  let distanceAdded = 0;
  if (lastPoint) {
    distanceAdded = haversineKm(lastPoint.lat, lastPoint.lng, lat, lng);
  }
  // insert new point
  const point: TripPointRow = await new Promise((resolve, reject) => {
    const stmt = db.prepare(
      "INSERT INTO trip_points (trip_id, lat, lng) VALUES (?, ?, ?)"
    );
    stmt.run(tripId, lat, lng, function (this: any, err: Error | null) {
      if (err) return reject(err);
      db.get(
        "SELECT * FROM trip_points WHERE id = ?",
        [this.lastID],
        (e, row) => {
          if (e) return reject(e);
          resolve(row as TripPointRow);
        }
      );
    });
  });
  if (distanceAdded > 0) {
    await new Promise((resolve, reject) => {
      db.run(
        "UPDATE trips SET total_distance_km = total_distance_km + ? WHERE id = ?",
        [distanceAdded, tripId],
        (err) => (err ? reject(err) : resolve(undefined))
      );
    });
  }
  const trip: TripRow = await new Promise((resolve, reject) => {
    db.get("SELECT * FROM trips WHERE id = ?", [tripId], (err, row) => {
      if (err) return reject(err);
      resolve(row as TripRow);
    });
  });
  return { point, distanceAdded, total: trip.total_distance_km };
}

export function stopTrip(tripId: number, finalFuel?: number): Promise<TripRow> {
  return new Promise((resolve, reject) => {
    db.run(
      "UPDATE trips SET ended_at = CURRENT_TIMESTAMP, final_fuel_liters = ? WHERE id = ? AND ended_at IS NULL",
      [finalFuel ?? null, tripId],
      (err) => {
        if (err) return reject(err);
        db.get("SELECT * FROM trips WHERE id = ?", [tripId], (e, row) => {
          if (e) return reject(e);
          resolve(row as TripRow);
        });
      }
    );
  });
}

export function listTripPoints(
  tripId: number,
  limit = 100
): Promise<TripPointRow[]> {
  return new Promise((resolve, reject) => {
    db.all(
      "SELECT * FROM trip_points WHERE trip_id = ? ORDER BY timestamp DESC LIMIT ?",
      [tripId, limit],
      (err, rows) => {
        if (err) return reject(err);
        resolve(rows as TripPointRow[]);
      }
    );
  });
}

export function getAllTrips(userId: number): Promise<TripRow[]> {
  return new Promise((resolve, reject) => {
    db.all(
      "SELECT * FROM trips WHERE user_id = ? ORDER BY started_at DESC",
      [userId],
      (err, rows) => {
        if (err) return reject(err);
        resolve(rows as TripRow[]);
      }
    );
  });
}

export function convertTripsToCSV(trips: TripRow[]): string {
  const headers = [
    "id",
    "user_id",
    "started_at",
    "ended_at",
    "initial_fuel_liters",
    "final_fuel_liters",
    "total_distance_km"
  ];
  
  const csvRows = [headers.join(",")];
  
  for (const trip of trips) {
    const row = [
      trip.id,
      trip.user_id,
      trip.started_at,
      trip.ended_at || "",
      trip.initial_fuel_liters ?? "",
      trip.final_fuel_liters ?? "",
      trip.total_distance_km
    ];
    csvRows.push(row.join(","));
  }
  
  return csvRows.join("\n");
}

File: server_python/app/trips.py
from typing import Optional, List, Tuple
import aiosqlite
from .models import Trip, TripPoint
from .calc import haversine_km


def row_to_trip(row) -> Trip:
    """Convert a database row to a Trip model."""
    return Trip(
        id=row["id"],
        user_id=row["user_id"],
        started_at=row["started_at"],
        ended_at=row["ended_at"],
        initial_fuel_liters=row["initial_fuel_liters"],
        final_fuel_liters=row["final_fuel_liters"],
        total_distance_km=row["total_distance_km"] or 0,
    )


def row_to_trip_point(row) -> TripPoint:
    """Convert a database row to a TripPoint model."""
    return TripPoint(
        id=row["id"],
        trip_id=row["trip_id"],
        timestamp=row["timestamp"],
        lat=row["lat"],
        lng=row["lng"],
    )


async def get_active_trip(db: aiosqlite.Connection, user_id: int) -> Optional[Trip]:
    """Get the active trip for a user."""
    async with db.execute(
        "SELECT * FROM trips WHERE user_id = ? AND ended_at IS NULL ORDER BY started_at DESC LIMIT 1",
        (user_id,)
    ) as cursor:
        row = await cursor.fetchone()
        return row_to_trip(row) if row else None


async def start_trip(db: aiosqlite.Connection, user_id: int, initial_fuel: Optional[float] = None) -> Trip:
    """Start a new trip for a user."""
    cursor = await db.execute(
        "INSERT INTO trips (user_id, initial_fuel_liters) VALUES (?, ?)",
        (user_id, initial_fuel)
    )
    await db.commit()
    
    async with db.execute(
        "SELECT * FROM trips WHERE id = ?",
        (cursor.lastrowid,)
    ) as cursor:
        row = await cursor.fetchone()
        return row_to_trip(row)


async def add_point(
    db: aiosqlite.Connection, 
    trip_id: int, 
    lat: float, 
    lng: float
) -> Tuple[TripPoint, float, float]:
    """Add a point to a trip and return the point, distance added, and total distance."""
    # Get last point
    async with db.execute(
        "SELECT * FROM trip_points WHERE trip_id = ? ORDER BY timestamp DESC LIMIT 1",
        (trip_id,)
    ) as cursor:
        last_point_row = await cursor.fetchone()
    
    distance_added = 0.0
    if last_point_row:
        last_point = row_to_trip_point(last_point_row)
        distance_added = haversine_km(last_point.lat, last_point.lng, lat, lng)
    
    # Insert new point
    cursor = await db.execute(
        "INSERT INTO trip_points (trip_id, lat, lng) VALUES (?, ?, ?)",
        (trip_id, lat, lng)
    )
    await db.commit()
    
    # Get the inserted point
    async with db.execute(
        "SELECT * FROM trip_points WHERE id = ?",
        (cursor.lastrowid,)
    ) as cursor:
        point_row = await cursor.fetchone()
        point = row_to_trip_point(point_row)
    
    # Update total distance if needed
    if distance_added > 0:
        await db.execute(
            "UPDATE trips SET total_distance_km = total_distance_km + ? WHERE id = ?",
            (distance_added, trip_id)
        )
        await db.commit()
    
    # Get updated trip
    async with db.execute(
        "SELECT * FROM trips WHERE id = ?",
        (trip_id,)
    ) as cursor:
        trip_row = await cursor.fetchone()
        total = trip_row["total_distance_km"] or 0
    
    return point, distance_added, total


async def stop_trip(db: aiosqlite.Connection, trip_id: int, final_fuel: Optional[float] = None) -> Trip:
    """Stop a trip and return the updated trip."""
    await db.execute(
        "UPDATE trips SET ended_at = CURRENT_TIMESTAMP, final_fuel_liters = ? WHERE id = ? AND ended_at IS NULL",
        (final_fuel, trip_id)
    )
    await db.commit()
    
    async with db.execute(
        "SELECT * FROM trips WHERE id = ?",
        (trip_id,)
    ) as cursor:
        row = await cursor.fetchone()
        return row_to_trip(row)


async def list_trip_points(db: aiosqlite.Connection, trip_id: int, limit: int = 100) -> List[TripPoint]:
    """List points for a trip."""
    async with db.execute(
        "SELECT * FROM trip_points WHERE trip_id = ? ORDER BY timestamp DESC LIMIT ?",
        (trip_id, limit)
    ) as cursor:
        rows = await cursor.fetchall()
        return [row_to_trip_point(row) for row in rows]


async def get_all_trips(db: aiosqlite.Connection, user_id: int) -> List[Trip]:
    """Get all trips for a user."""
    async with db.execute(
        "SELECT * FROM trips WHERE user_id = ? ORDER BY started_at DESC",
        (user_id,)
    ) as cursor:
        rows = await cursor.fetchall()
        return [row_to_trip(row) for row in rows]


def convert_trips_to_csv(trips: List[Trip]) -> str:
    """Convert a list of trips to CSV format."""
    headers = [
        "id",
        "user_id",
        "started_at",
        "ended_at",
        "initial_fuel_liters",
        "final_fuel_liters",
        "total_distance_km"
    ]
    
    csv_rows = [",".join(headers)]
    
    for trip in trips:
        row = [
            str(trip.id),
            str(trip.user_id),
            trip.started_at,
            trip.ended_at or "",
            str(trip.initial_fuel_liters) if trip.initial_fuel_liters is not None else "",
            str(trip.final_fuel_liters) if trip.final_fuel_liters is not None else "",
            str(trip.total_distance_km)
        ]
        csv_rows.append(",".join(row))
    
    return "\n".join(csv_rows)
